modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/PaginationContextProvider.java

```diff
@@ -22,21 +32,74 @@
 @Provider
 public class PaginationContextProvider implements ContextProvider<Pagination> {
 
+	public PaginationContextProvider(
+		ConfigurationProvider configurationProvider, Portal portal) {
+
+		_configurationProvider = configurationProvider;
+		_portal = portal;
+	}
+
 	@Override
 	public Pagination createContext(Message message) {
 		HttpServletRequest httpServletRequest =
 			ContextProviderUtil.getHttpServletRequest(message);
 
-		String page = httpServletRequest.getParameter("page");
-		String pageSize = httpServletRequest.getParameter("pageSize");
+		int page = GetterUtil.getInteger(
+			httpServletRequest.getParameter("page"), 1);
 
-		if (StringUtil.equals(page, "0") || StringUtil.equals(pageSize, "0")) {
-			return null;
+		if (page <= 0) {
+			throw new InvalidPaginationException(
+				String.format(
+					"The Page parameter introduced [%s] is not valid. Only " +
+						"numbers higher or equal to 1 are accepted.",
+					page));
 		}
 
-		return Pagination.of(
-			GetterUtil.getInteger(page, 1),
-			GetterUtil.getInteger(pageSize, 20));
+		int pageSize = GetterUtil.getInteger(
+			httpServletRequest.getParameter("pageSize"), 20);
+
+		try {
+			HeadlessAPICompanyConfiguration headlessAPICompanyConfiguration =
+				_configurationProvider.getCompanyConfiguration(
+					HeadlessAPICompanyConfiguration.class,
+					_portal.getCompanyId(httpServletRequest));
+
+			int maxPageSize =
+				headlessAPICompanyConfiguration.paginationSizeLimit();
+
+			if (maxPageSize <= 0) {
+				return Pagination.of(
+					page, (pageSize <= 0) ? QueryUtil.ALL_POS : pageSize);
+			}
+
+			if ((pageSize > maxPageSize) || (pageSize <= 0)) {
```

Comment:
I think this condition must be split in 2:
1. If `pageSize == -1` then apply the same logic to avoid breaking changes.
2. If `pageSize > 0` then throw an error informing the user about the limit. If we don't throw the error the user will get unexpected results. If they ask for `pageSize=500` and `page=2` they would expect to see the `501` element as first result. With this code they would get the `301` without any warning at all.

Also, I would remove the `_log.warn` because this warning is irrelevant for the system admin. The API user is who should be notified.

Reply:
After discussing on a call, we will keep the current behaviour. For now remove the log to avoid attackers filling it, and add a message in system settings to explain what will happen in these 2 cases.